topdown/providers.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"encoding/json"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/internal/providers"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+var awsRequiredConfigKeyNames = ast.NewSet(
+	ast.StringTerm("aws_service"),
+	ast.StringTerm("aws_access_key"),
+	ast.StringTerm("aws_secret_access_key"),
+	ast.StringTerm("aws_region"),
+)
+
+func stringFromTerm(t *ast.Term) string {
+	if v, ok := t.Value.(ast.String); ok {
+		return string(v)
+	}
+	return ""
+}
+
+func getReqBodyBytes(body, rawBody *ast.Term) ([]byte, error) {
+	var out []byte
+
+	switch {
+	case rawBody != nil:
+		out = []byte(stringFromTerm(rawBody))
+	case body != nil:
+		bodyVal := body.Value
+		bodyValInterface, err := ast.JSON(bodyVal)
+		if err != nil {
+			return nil, err
+		}
+		bodyValBytes, err := json.Marshal(bodyValInterface)
+		if err != nil {
+			return nil, err
+		}
+		out = bodyValBytes
+	}
+
+	if out == nil {
```

Comment:
Could this be true after any of the `case` branches above? If not, let's put it into `default:`

Reply:
Good suggestion; it has been rolled into the latest batch of changes! :+1: